File: LCA.scala
lass Entity

class Object

class Property

class Relation(val subject: Entity, val obj: Object, val properties: List[Property])

class Correspondence(val mapping: List[List[Double]], val properties: List[Property])

class LinearCorrespondence(val correspondence: Correspondence) {
  def isLinear(prop1: Property, prop2: Property, rel1: Relation, rel2: Relation): Boolean = {
    for (r1 <- rel1.properties) {
      for (r2 <- rel2.properties) {
        if (rel1.subject eq rel2.subject) && (rel1.obj ne rel2.obj) {
          val prop1_ij = correspondence.mapping(properties.indexOf(prop1))(rel1.properties.indexOf(r1))
          val prop1_jk = correspondence.mapping(properties.indexOf(prop1))(rel2.properties.indexOf(r2))
          val prop2_ij = correspondence.mapping(properties.indexOf(prop2))(rel1.properties.indexOf(r1))
          val prop2_jk = correspondence.mapping(properties.indexOf(prop2))(rel2.properties.indexOf(r2))
          if (prop1_ij != prop1_jk || prop2_ij != prop2_jk) {
            return false
          }
        }
      }
    }
    true
  }
}

//Example usage
val john = new Entity
val mary = new Entity
val book = new Object

val color = new Property
val size = new Property

val on1 = new Relation(john, book, List(color, size))
val on2 = new Relation(mary, book, List(color, size))

val color_size_corr = new Correspondence(List(List(0.0, 0.8), List(0.2, 0.0)), List(color, size))
val size_color_corr = new Correspondence(List(List(0.0, 0.6), List(0.4, 0.0)), List(size, color))
val color_size_lin_corr = new LinearCorrespondence(color_size_corr)

color_size_lin_corr.isLinear(color, size, on1, on2) // returns true

size_color_lin_corr.isLinear(color, size, on1, on2) // returns false

File: LCA.go
package main

import (
	"fmt"
)

// Define entities
type Entity struct{}

// Define objects
type Object struct{}

// Define properties
type Property struct{}

// Define relations
type Relation struct {
	subject Entity
	object  Object
	props   []Property
}

// Define correspondences
type Correspondence struct {
	mapping map[Property]map[Entity]float64
}

// Define linear correspondences
type LinearCorrespondence struct {
	Correspondence
}

// Define function to check if a pair of relations satisfy the Linear Correspondence Axiom
func isLinearCorrespondence(prop1 Property, prop2 Property, rel1 Relation, rel2 Relation, corr Correspondence) bool {
	isLinear := true
	for _, r1 := range rel1.props {
		for _, r2 := range rel2.props {
			if rel1.subject == rel2.subject && rel1.object != rel2.object {
				prop1_ij := corr.mapping[prop1][r1]
				prop1_jk := corr.mapping[prop1][r2]
				prop2_ij := corr.mapping[prop2][r1]
				prop2_jk := corr.mapping[prop2][r2]
				if !(prop1_ij == prop1_jk && prop2_ij == prop2_jk) {
					isLinear = false
					return isLinear
				}
			}
		}
	}
	return isLinear
}

func main() {
	// Define example entities
	john := Entity{}
	mary := Entity{}
	book := Object{}

	// Define example properties
	color := Property{}
	size := Property{}

	// Define example relations
	on1 := Relation{subject: john, object: book, props: []Property{color, size}}
	on2 := Relation{subject: mary, object: book, props: []Property{color, size}}

	// Define example correspondences
	colorSizeCorr := Correspondence{
		mapping: map[Property]map[Entity]float64{
			color: map[Entity]float64{
				size: 0.8,
			},
			size: map[Entity]float64{
				color: 0.2,
			},
		},
	}
	sizeColorCorr := Correspondence{
		mapping: map[Property]map[Entity]float64{
			color: map[Entity]float64{
				size: 0.6,
			},
			size: map[Entity]float64{
				color: 0.4,
			},
		},
	}
	colorSizeLinCorr := LinearCorrespondence{colorSizeCorr}
	sizeColorLinCorr := LinearCorrespondence{sizeColorCorr}

	// Check if on1 and on2 satisfy the Linear CorrespondenceAxiom
	if isLinearCorrespondence(color, size, on1, on2, colorSizeLinCorr) && isLinearCorrespondence(size, color, on1, on2, sizeColorLinCorr) {
		fmt.Println("The Linear Correspondence Axiom is satisfied.")
	} else {
		fmt.Println("The Linear Correspondence Axiom is not satisfied.")
	}
}
